plugins/inputs/http_listener/http_listener.go

```diff
@@ -111,36 +133,112 @@ func (t *HttpListener) httpListen() error {
+			return
 		}
 
-		if err := scanner.Err(); err != nil {
-			http.Error(res, "Internal server error: "+err.Error(), http.StatusInternalServerError)
-		} else if http400msg.Len() > 0 {
-			res.Header().Set("Content-Type", "application/json")
-			res.Header().Set("X-Influxdb-Version", "1.0")
-			res.WriteHeader(http.StatusBadRequest)
-			res.Write([]byte(fmt.Sprintf(`{"error":"%s%s"}`, partial, http400msg.String())))
+		// Handle gzip request bodies
+		var body io.ReadCloser
+		var err error
+		if req.Header.Get("Content-Encoding") == "gzip" {
+			body, err = gzip.NewReader(http.MaxBytesReader(res, req.Body, t.MaxBodySize))
+			if err != nil {
+				msg400.WriteString(err.Error() + " ")
+				return
+			}
 		} else {
-			res.WriteHeader(http.StatusNoContent)
+			body = http.MaxBytesReader(res, req.Body, t.MaxBodySize)
+		}
+
+		var buffer *bytes.Buffer
+		if req.ContentLength < MAX_ALLOCATION_SIZE {
+			// if the content length is less than the max allocation size, then
+			// read in the whole request at once:
+			buffer = bytes.NewBuffer(make([]byte, 0, req.ContentLength+1))
+			_, err := buffer.ReadFrom(body)
+			if err != nil {
+				msg := "E! "
+				if netErr, ok := err.(net.Error); ok {
+					if netErr.Timeout() {
+						msg += "Read timeout error, you may want to increase the read_timeout setting. "
+					}
+				}
+				log.Printf(msg + err.Error())
+				msg400.WriteString("Error reading request body: " + err.Error())
+				return
+			}
+		} else {
+			// If the body is larger than the max allocation size then set the
+			// maximum size of the buffer that we will allocate at a time.
+			// The following loop goes through the request body byte-by-byte.
+			// If there is a newline within 256 kilobytes of the end of the body
+			// we will attempt to parse metrics, reset the buffer, and continue.
+			buffer = bytes.NewBuffer(make([]byte, 0, MAX_ALLOCATION_SIZE))
+			reader := bufio.NewReader(body)
+			for {
+				b, err := reader.ReadByte()
+				if err != nil {
+					if err != io.EOF {
+						msg := "E! "
+						if netErr, ok := err.(net.Error); ok {
+							if netErr.Timeout() {
+								msg += "Read timeout error, you may want to increase the read_timeout setting. "
+							}
+						} else {
+							// if it's not an EOF or a net.Error, then it's almost certainly a
+							// tooLarge error coming from http.MaxBytesReader. It's unlikely
+							// that this code path will get hit because the client should
+							// be setting the ContentLength header, unless it's malicious.
+							msg413.WriteString(err.Error())
+						}
+						log.Printf(msg + err.Error())
+						return
+					}
+					break
+				}
+				// returned error is always nil:
+				// https://golang.org/pkg/bytes/#Buffer.WriteByte
+				buffer.WriteByte(b)
+				// if we have a newline and we're nearing the end of the buffer,
+				// do a write and continue with a fresh buffer.
+				if buffer.Len() > MAX_ALLOCATION_SIZE-256*1000 && b == '\n' {
+					t.parse(buffer.Bytes(), &msg400)
```

Comment:
This chunked processing differs in behavior from the standard influxdb handler in that these "chunks" will end up with separate timestamps, if they weren't set on the original points. If there's going to be chunked processing, the current time should be grabbed at the beginning of the processing, then passed around as the default time to `models.ParsePointsWithPrecision`. I see that `models.ParsePoints` is used in the parser, which just [calls](https://github.com/influxdata/influxdb/blob/8fce6bba4814df34017117ba97270791520936af/models/points.go#L178) `models.ParsePointsWithPrecision(buf, time.Now(), "n")`. Capturing the timestamp earlier and replacing all uses of `ParsePoints(buf)` with `ParsePointsWithPrecision(buf, ts, "n")` would keep similar behavior to the database handler.